orderer/common/follower/follower_chain.go

```diff
@@ -286,146 +255,277 @@ func (c *Chain) run() {
+	if *retryInterval == lowerLimit {
+		return
 	}
-	c.logger.Debugf("retry interval increased to: %v", c.retryInterval)
+	*retryInterval = lowerLimit
+	c.logger.Debugf("retry interval reset to: %v", *retryInterval)
 }
 
-func (c *Chain) resetRetryInterval() {
-	if c.retryInterval == c.options.PullRetryMinInterval {
+func (c *Chain) decreaseRetryInterval(retryInterval *time.Duration, lowerLimit time.Duration) {
+	if *retryInterval == lowerLimit {
 		return
 	}
-	c.retryInterval = c.options.PullRetryMinInterval
-	c.logger.Debugf("retry interval reset to: %v", c.retryInterval)
+
+	*retryInterval = *retryInterval - lowerLimit
+	if *retryInterval < lowerLimit {
+		*retryInterval = lowerLimit
+	}
+	c.logger.Debugf("retry interval decreased to: %v", *retryInterval)
 }
 
-// pull blocks from other orderers, until the join block, or until a config block the indicates the orderer has become
-// a member pf the cluster. When the follower.Chain's job is done, this method halts, triggers the creation of a new
-// consensus.Chain, and returns nil.
+// pull blocks from other orderers until a config block indicates the orderer has become a member of the cluster.
+// When the follower.Chain's job is done, this method halts, triggers the creation of a new consensus.Chain,
+// and returns nil. The method returns an error only when the chain is stopped or due to unrecoverable errors.
 func (c *Chain) pull() error {
+	var err error
 	if c.joinBlock != nil {
-		err := c.pullUpToJoin()
+		err = c.pullUpToJoin()
 		if err != nil {
 			return errors.Wrap(err, "failed to pull up to join block")
 		}
 
 		// TODO remove the join-block from the file repo
-	} else {
-		err := c.pullAfterJoin()
-		if err != nil {
-			return errors.Wrap(err, "failed to pull after join block")
-		}
-	}
-
-	// Trigger creation of a new chain. This will regenerate 'ledgerResources' from the tip of the ledger, and will
-	// start either a follower.Chain or an etcdraft.Chain, depending on whether the orderer is a follower or a
-	// member of the cluster.
-	c.halt()
 
-	if c.joinBlock != nil {
+		// The join block never returns an error. This is checked before the follower is started.
 		if isMem, _ := c.clusterConsenter.IsChannelMember(c.joinBlock); isMem {
+			c.setStatusReport(types.ClusterRelationMember, types.StatusActive)
+			// Trigger creation of a new consensus.Chain.
 			c.logger.Info("On-boarding finished successfully, invoking chain creation callback")
+			c.halt()
 			c.chainCreationCallback()
-		} else {
-			c.logger.Info("On-boarding finished successfully, invoking follower creation callback")
-			c.followerCreationCallback()
+			return nil
 		}
-	} else {
-		c.logger.Info("Block pulling finished successfully, invoking chain creation callback")
-		c.chainCreationCallback()
+
+		c.setStatusReport(types.ClusterRelationFollower, types.StatusActive)
+		c.joinBlock = nil
+		err = c.loadLastConfig()
```

Comment:
config is needed on entry to the pullAfterJoin method. It is initialized in the constructor in case we start the follower without a join block (say after restart) and then the only other place it needs to be set is during  the transition from until-join to after-join.